crates/ruff_db/src/parsed.rs

```diff
@@ -0,0 +1,99 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_python_ast::{ModModule, PySourceType};
+use ruff_python_parser::{parse_unchecked_source, Parsed};
+
+use crate::source::source_text;
+use crate::vfs::{VfsFile, VfsPath};
+use crate::Db;
+
+/// Returns the parsed AST of `file`, including its token stream.
+///
+/// The query uses Ruff's error resilient parser. That means that the parser always succeeds to produce a
+/// AST even if the file contains syntax errors. The syntax errors are Parsing the module succeeds even when the file contains syntax error. The parse errors
+/// are then accessible through [`Parsed::errors`].
+///
+/// The parse tree is cached between invocations, but the query doesn't make use of Salsa's optimization
+/// that skips dependent queries if the AST hasn't changed. Comparing two ASTs is a non-trivial operation
+/// and every offset change is directly reflected in the changed AST offsets. Ruff's AST also doesn't implement `Eq`.
+/// which is required to use the optimization.
+#[salsa::tracked(return_ref, no_eq)]
+pub fn parsed_module(db: &dyn Db, file: VfsFile) -> Parsed<ModModule> {
+    let source = source_text(db, file);
+    let path = file.path(db);
+
+    let ty = match path {
+        VfsPath::FileSystem(path) => path
+            .extension()
+            .map_or(PySourceType::Python, PySourceType::from_extension),
```

Comment:
Does it make sense to assume a file is Python if it has no extension? Is that an existing Ruff behavior? That seems likely to lead to a lot of syntax/parsing errors. From a user perspective, I would expect to have to specially request via config if I want Ruff to treat an unknown-extension file as Python source.

Reply:
We'll need to filter out non-python files (according to the user's configuration) when walking the workspace so that we never end up calling parse on these files.